src/nv_ingest/util/pdf/pdfium.py

```diff
@@ -188,3 +195,354 @@ def pdfium_pages_to_numpy(
         images.append(img_arr)
 
     return images, padding_offsets
+
+
+def convert_pdfium_position(pos, page_width, page_height):
+    """
+    Convert a PDFium bounding box (which typically has an origin at the bottom-left)
+    to a more standard bounding-box format with y=0 at the top.
+    """
+    left, bottom, right, top = pos
+    x0, x1 = left, right
+    y0, y1 = page_height - top, page_height - bottom
+
+    x0 = max(0, x0)
+    y0 = max(0, y0)
+    x1 = min(page_width, x1)
+    y1 = min(page_height, y1)
+
+    return [int(x0), int(y0), int(x1), int(y1)]
+
+
+def boxes_are_close_or_overlap(b1, b2, threshold=10.0):
+    """
+    Return True if bounding boxes b1 and b2 overlap or
+    are closer than 'threshold' in points/pixels.
+    b1, b2 = (xmin, ymin, xmax, ymax).
+    """
+    (xmin1, ymin1, xmax1, ymax1) = b1
+    (xmin2, ymin2, xmax2, ymax2) = b2
+
+    # Check if they overlap horizontally
+    overlap_x = not (xmax1 < xmin2 or xmax2 < xmin1)
+    # Check if they overlap vertically
+    overlap_y = not (ymax1 < ymin2 or ymax2 < ymin1)
+
+    # If there's an actual overlap area, that's enough
+    if overlap_x and overlap_y:
+        return True
+
+    # Otherwise, measure the gap. We can do a simple approach:
+    #   horizontal gap = distance between the rightmost left edge and the leftmost right edge
+    #   vertical gap   = ...
+    # But to keep it straightforward, let's do a quick bounding box expansions approach
+    # Expand each box by 'threshold' in all directions and see if they overlap now
+    expanded_b1 = (xmin1 - threshold, ymin1 - threshold, xmax1 + threshold, ymax1 + threshold)
+    expanded_b2 = (xmin2 - threshold, ymin2 - threshold, xmax2 + threshold, ymax2 + threshold)
+
+    # Check overlap on expanded boxes
+    (exmin1, eymin1, exmax1, eymax1) = expanded_b1
+    (exmin2, eymin2, exmax2, eymax2) = expanded_b2
+
+    overlap_x_expanded = not (exmax1 < exmin2 or exmax2 < exmin1)
+    overlap_y_expanded = not (eymax1 < eymin2 or eymax2 < eymin1)
+
+    return overlap_x_expanded and overlap_y_expanded
+
+
+def group_bounding_boxes(boxes, threshold=10.0, max_depth=None):
+    """
+    Given a list of bounding boxes,
+    returns a list of groups (lists) of bounding box indices.
+    """
+    n = len(boxes)
+    visited = [False] * n
+    adjacency_list = [[] for _ in range(n)]
+
+    # Build adjacency by checking closeness/overlap
+    for i in range(n):
```

Comment:
This is probably fine for small numbers of boxes; but its still O(n^2), which could cause problems on some of the pathological PDFs we've seen in the past. Some of the US-GS topo pdfs have layouts where there is effectively 1 image per pixel, and 10s of 1000s of images on the page.

I may be missing if you're doing this somewhere, but I think we should skip grouping or emit some warnings if the number of bounding boxes on a page exceeds some threshold. If we exceed that threshold maybe switch to something like a quadtree for grouping.

Reply:
I think it makes sense to skip grouping and emit a warning if the number of bounding boxes exceed the threshold. From my experiments, most of the gain come from using the bounding box of PdfFormObject and grouping PdfPathObject. These PDFs with an unreasonable number of objects don't gain much by grouping them because they seem to be just badly constructed PDFs to begin with. In my opinion, they would be better solved by OCR models.